File: Python/089__Gray_Code.py
class Solution(object):
    def grayCode(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        if n == 0:
            return [0]
        elif n == 1:
            return [0, 1]
        else:
            l1 = self.grayCode(n - 1)
            l2 = [l + (1 << (n - 1)) for l in l1[::-1]]
        return l1 + l2

File: Java/089__Gray_Code.java
class Solution {
    public List<Integer> grayCode(int n) {
        int len = (int) Math.pow(2, n);
        List<Integer> result = new ArrayList<>(len);
        for (int i = 0; i < len; i++)
            result.add(i ^ (i >> 1));
        return result;
    }
}
